auth/validator.go

```diff
@@ -5,119 +5,128 @@
 package auth
 
 import (
+	"errors"
 	"fmt"
-	"github.com/dgrijalva/jwt-go/v4"
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/lestrrat-go/jwx/jws"
+	"github.com/lestrrat-go/jwx/jwt"
 	"github.com/sap/cloud-security-client-go/oidcclient"
 	"net/url"
 	"strings"
 	"time"
 )
 
 // parseAndValidateJWT parses the token into its claims, verifies the claims and verifies the signature
-func (m *Middleware) parseAndValidateJWT(rawToken string) (*jwt.Token, error) {
-	token, parts, err := m.parser.ParseUnverified(rawToken, new(OIDCClaims))
+func (m *Middleware) parseAndValidateJWT(rawToken string) (Token, error) {
+
+	token, err := NewToken(rawToken)
 	if err != nil {
 		return nil, err
 	}
-	token.Signature = parts[2]
 
 	// get keyset
-	keySet, err := m.getOIDCTenant(token)
+	keySet, err := m.getOIDCTenant(token.Issuer())
 	if err != nil {
 		return nil, err
 	}
 
-	// verify claims
-	if err := m.validateClaims(token, keySet); err != nil {
-		return nil, err
-	}
-
 	// verify signature
-	if err = m.verifySignature(token, keySet); err != nil {
+	if err := m.verifySignature(token, *keySet); err != nil {
 		return nil, err
 	}
 
-	mapClaims, _, err := m.parser.ParseUnverified(rawToken, jwt.MapClaims{})
-	if err != nil {
+	// verify claims
+	if err := m.validateClaims(token, keySet); err != nil {
 		return nil, err
 	}
-	token.Claims.(*OIDCClaims).mapClaims = mapClaims.Claims.(jwt.MapClaims)
 
-	token.Valid = true
 	return token, nil
 }
 
-func (m *Middleware) verifySignature(t *jwt.Token, ks *oidcclient.OIDCTenant) error {
-	jwks, err := ks.GetJWKs()
+func (m *Middleware) verifySignature(t Token, keySet oidcclient.OIDCTenant) (err error) {
+	headers, err := getHeaders(t.GetTokenValue())
 	if err != nil {
-		return fmt.Errorf("token is unverifiable: failed to fetch token keys from remote: %v", err)
+		return err
 	}
-	if len(jwks) == 0 {
-		return fmt.Errorf("token is unverifiable: remote returned no jwk to verify the token")
+	kid := headers.KeyID()
+	alg := headers.Algorithm()
+
+	if alg == "" {
+		return errors.New("alg is missing from jwt header")
```

Comment:
this check is not required - jwt.WithVerify() fails in case alg is empty

Reply:
but I want to avoid parsing again, if the alg is missing